Fix sym_{sizes,strides} slow path

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107089
* __->__ #107839

Previously, when SymInt is returned from sym_sizes slow path, it would segfault.

This is useful for tensors that have symbolic sizes and use the sym_sizes slow path, e.g. NestedTensor returning SingletonSymInt as its sizes in the slow path.

See also: https://github.com/pytorch/pytorch/pull/106405/files#r1303714865

soulitzer committed:
Fix sym_{sizes,strides} slow path

[ghstack-poisoned]

ezyang commented on torch/csrc/PyInterpreter.cpp:
```diff
@@ -587,28 +587,27 @@ c10::IntArrayRef ConcretePyInterpreterVTable::strides(
+    const c10::TensorImpl* tensor,
+    T* array_data,
+    const char* attr_name) {
+  c10::TensorImpl* t_ptr = const_cast<c10::TensorImpl*>(tensor);
```
Why not just take a non-const pointer as argument?

ezyang commented on torch/csrc/PyInterpreter.cpp:
```diff
@@ -587,28 +587,27 @@ c10::IntArrayRef ConcretePyInterpreterVTable::strides(
-  auto result = THPUtils_unpackLongs(buffer.ptr());
-  return result;
+  auto capsule = py::capsule(array_data, [](void* p) {
+      delete[] reinterpret_cast<T*>(p);
```
This is suspicious to me. It is better code structure if the deleter is created at the same spot where the data is allocated, so it is easier to audit that you have lined them up correctly.

ezyang reviewed (approved):
This seems functionally correct

soulitzer committed:
Update on "Fix sym_{sizes,strides} slow path"


This is useful for tensors that have symbolic sizes and use the sym_sizes slow path, e.g. NestedTensor returning SingletonSymInt as its sizes in the slow path.

See also: https://github.com/pytorch/pytorch/pull/106405/files#r1303714865

[ghstack-poisoned]

soulitzer committed:
Update on "Fix sym_{sizes,strides} slow path"


This is useful for tensors that have symbolic sizes and use the sym_sizes slow path, e.g. NestedTensor returning SingletonSymInt as its sizes in the slow path.

See also: https://github.com/pytorch/pytorch/pull/106405/files#r1303714865

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge
